internal/aws/xray/telemetry/recorder.go

```diff
@@ -0,0 +1,327 @@
+		SegmentsRejectedCount:  aws.Int64(0),
+		SegmentsSentCount:      aws.Int64(0),
+		SegmentsSpilloverCount: aws.Int64(0),
+		BackendConnectionErrors: &xray.BackendConnectionErrors{
+			HTTPCode4XXCount:       aws.Int64(0),
+			HTTPCode5XXCount:       aws.Int64(0),
+			ConnectionRefusedCount: aws.Int64(0),
+			OtherCount:             aws.Int64(0),
+			TimeoutCount:           aws.Int64(0),
+			UnknownHostCount:       aws.Int64(0),
+		},
+	}
+}
+
+// Start starts the loop to send the records.
+func (t *telemetryRecorder) Start() {
+	t.startOnce.Do(func() {
+		go t.start()
+	})
+}
+
+// Stop closes the done channel to stop the loop.
+func (t *telemetryRecorder) Stop() {
+	t.stopOnce.Do(func() {
+		close(t.done)
+	})
+}
+
+// start flushes the record once a minute if telemetry
+// data was recordUpdated.
+func (t *telemetryRecorder) start() {
+	ticker := time.NewTicker(t.interval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			if t.recordUpdated.Load() {
+				t.add(t.cutoff())
+				t.recordUpdated.Store(false)
+				if failedToSend, err := t.send(t.flush()); err != nil {
+					for _, record := range failedToSend {
+						t.add(record)
+					}
+				}
+			}
+		case <-t.done:
+			return
+		}
+	}
+}
+
+// cutoff the current record and swap it out with a new record.
+// Sets the timestamp and returns the old record.
+func (t *telemetryRecorder) cutoff() *xray.TelemetryRecord {
+	t.mu.Lock()
+	oldRecord := t.record
+	t.record = newTelemetryRecord()
+	t.mu.Unlock()
+	oldRecord.SetTimestamp(time.Now())
+	return oldRecord
+}
+
+// add to the queue. If queue is full, drop the head of the queue and try again.
+func (t *telemetryRecorder) add(record *xray.TelemetryRecord) {
+	select {
+	case t.queue <- record:
+	default:
+		<-t.queue
+		t.add(record)
+	}
+}
+
+// flush the queue into a slice of records.
+func (t *telemetryRecorder) flush() []*xray.TelemetryRecord {
+	var records []*xray.TelemetryRecord
+	for len(t.queue) > 0 {
+		records = append(records, <-t.queue)
+	}
+	return records
+}
+
+// send the records in batches of batchSize. Returns the error and records it was unable to send
+// if the PutTelemetryRecords call fails.
+func (t *telemetryRecorder) send(records []*xray.TelemetryRecord) ([]*xray.TelemetryRecord, error) {
+	if len(records) > 0 {
+		for i := 0; i < len(records); i += batchSize {
+			endIndex := i + batchSize
+			if endIndex > len(records) {
+				endIndex = len(records)
+			}
+			input := &xray.PutTelemetryRecordsInput{
+				EC2InstanceId:    &t.instanceID,
+				Hostname:         &t.hostname,
+				ResourceARN:      &t.resourceARN,
+				TelemetryRecords: records[i:endIndex],
+			}
+			_, err := t.client.PutTelemetryRecords(input)
+			if err != nil {
+				t.RecordConnectionError(err)
+				return records[i:], err
+			}
+		}
+	}
+	return nil, nil
+}
+
+func (t *telemetryRecorder) RecordSegmentsReceived(count int) {
+	t.mu.Lock()
+	atomic.AddInt64(t.record.SegmentsReceivedCount, int64(count))
```

Comment:
You shouldn't need a lock around an atomic value, it by definition is safe for concurrent access.

Reply:
The main problem I had with this is that I swap out the record in `cutoff`, so there's a datarace if this is called during that. It'll update the swapped out records instead of the new one.

Reply:
You could use [atomic.SwapInt64](https://pkg.go.dev/sync/atomic#SwapInt64) in order to perform this?

It would still provide the same ordering guarantees that you're currently getting with the Mutex + atomic

Reply:
You're right. I was trying to swap the entire struct out, but swapping the pointer fields makes more sense. Removed the mutex.